### PR TITLE
docs: implement versioned documentation

### PR DESCRIPTION
**Pull request by srdtrk**
### Overview
- [x] Docs for the following IBC-go versions are now stored in the `feat` (later `main`) branch:
  - [x] `v7.0.0`
  - [x] `v6.1.0`
  - [x] `v5.3.0`
  - [x] `v4.4.0`
- [x] The `link-check` workflow now supports docusaurus links ~~and is only activated upon `.md` file modifications. (excluding `.github/` directory)~~
- [ ] ~~New `markdown-lint` workflow has been added~~ (moved to another PR)
  - We've shifted from  [`markdownlint`](https://github.com/DavidAnson/markdownlint) to [`markdownlint-cli2`](https://github.com/DavidAnson/markdownlint-cli2) in this repo. (Makefile is updated accordingly)
  - This workflow only runs on modified markdown files if a `.md` file has been modified (not including `.github/` directory).
  - There is a new config file for this tool at [`.markdownlint-cli2.jsonc`](https://github.com/cosmos/ibc-go/pull/3577/files#diff-4929acf5ce7500ed4d3f00992a1c918e702427b7857fed53152e48fd58c2ae9c), this tool also continues to use [`.markdownlint.jsonc`](https://github.com/cosmos/ibc-go/blob/serdar/issue%233522-store-docs-for-past-versions/.markdownlint.jsonc)
  - Additional rules are added to [`.markdownlint.jsonc`](https://github.com/cosmos/ibc-go/blob/serdar/issue%233522-store-docs-for-past-versions/.markdownlint.jsonc) to address: #3543.
- [x] Makefile and docs deployment workflows have been updated
- [x] Added search bar to docs

closes: #3522, #3623

addresses: #3488 

#### Structural Changes

This update introduces versioned docs. Here, I will describe the structural changes that this feature brings, to learn more about versioning in this repo, go to [`docs/README.md`](https://github.com/cosmos/ibc-go/blob/serdar/issue%233522-store-docs-for-past-versions/docs/README.md#versioning).

Two new directories and one new file was added to `docs/` to enable versioning, 

```ignore
ibc-go/docs/
├── versions.json          # file to indicate what versions are available
├── versioned_docs/
│   ├── version-v4.4.0/    # markdown files and categories for v4.4.0 are stored here
│   ├── version-v5.3.0/
│   ├── version-v6.1.0/
│   └── version-v7.0.0/
├── versioned_sidebars/    # sidebars for the versions are kept here
│   ├── version-v4.4.0-sidebars.json
│   ├── version-v5.3.0-sidebars.json
│   ├── version-v6.1.0-sidebars.json
│   └── version-v7.0.0-sidebars.json
...
```

#### Content Changes:

- [x] ADRs' README.md is not missing some ADRs anymore.
- [x] Transfer is now the first IBC app.
- [x] README.md now also includes documentation guidelines. (DOCS_GUIDELINES.md no longer exists)
- [x] Fixes various styling errors and broken links.

### Commit Message / Changelog Entry

```
docs: implement versioned documentation and update workflows
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] ~Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).~
- [x] ~Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).~
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [x] ~Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).~
- [x] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] ~Re-reviewed `Files changed` in the Github PR explorer.~
- [x] ~Review `Codecov Report` in the comment section below once CI passes.~

### CODE REVIEW

File: .markdownlint.jsonc [hunk 1]
```diff
@@ -1,10 +1,17 @@
 {
   "default": true,
+  "MD003": { "style": "atx" },
```
Comment (DimitrisJim) [inline .markdownlint.jsonc:3]: can we also add a ref to the rule as is done with the others? I get most of these but it would still be good to keep a reference to them handily around.

**Comment by colin-axner** [comment]
Hi @srdtrk, is it possible to split up this pr? Smaller pr's would it make it easier to do reviews, catch issues, and merge code faster. With github's current review tooling, I think it's a big ask to have anyone from the team review a pr of this size and I suspect this pr will sit for a while longer if it isn't broken up

**Comment by srdtrk** [comment]
Yeah, you're right @colin-axner, I'd like to talk about how I might break it in the next planning call. Also, I'm curious if it's possible for me to walk over the PR in a call since its size is due to me copy pasting and styling a large amount of old docs but the main content of the PR is not that large imo.

**Comment by srdtrk** [comment]
I've removed the unnecessary linter changes. Let me know if this is easier to review @colin-axner

File: .github/ISSUE_TEMPLATE/release-tracker.md [hunk 2]
```diff
@@ -57,9 +57,14 @@ versions of ibc-go to guarantee that no regression is introduced -->
   - Add the new release.
   - Remove any tags that might not be recommended anymore.
 - [ ] Update the list of [supported release lines in README.md](https://github.com/cosmos/ibc-go#releases), if necessary.
-- [ ] (TODO: [#3522](https://github.com/cosmos/ibc-go/issues/3522)) Update docs site:
-  - [ ] Add new release branch to [`docs/versions`](https://github.com/cosmos/ibc-go/blob/main/docs/versions) file.
-  - [ ] Add `label` and `key` to `versions` array in [`config.js`](https://github.com/cosmos/ibc-go/blob/main/docs/.vuepress/config.js#L62).
+- [ ] Update docs site:
+  - [ ] If the release is occurring on the main branch, on the latest version, then run `npm run docusaurus docs:version vX.Y.Z` in the `docs/` directory. (where `X.Y.Z` is the new version number)
```
Comment (crodriguezvega) [inline .github/ISSUE_TEMPLATE/release-tracker.md:61]: We always release from a `release/vx.y.z` branch, never from main. Would this flow still work? Like, can I run `npm run docusaurus docs:version vX.Y.Z` in the `docs/` folder in `release/vX.Y.Z` branch?

File: .github/ISSUE_TEMPLATE/release-tracker.md [hunk 3]
```diff
@@ -57,9 +57,14 @@ versions of ibc-go to guarantee that no regression is introduced -->
...
+  - [ ] If the release is occurring on an older release branch, then make a PR to the main branch called `docs: new release vX.Y.Z` doing the following:
```
Comment (crodriguezvega) [inline .github/ISSUE_TEMPLATE/release-tracker.md:62]: `x.y.z` is here the previous (existing) release or the new one? Sorry, it was not clear to me...

File: docs/README.md [hunk 4]
```diff
@@ -84,19 +90,179 @@ slug: /migrations/v5-to-v6 # the url of the file
...
+- Either Leave a space or use a `-` between the acronyms ADR and ICS and the corresponding number (e.g. ADR 008 or ADR-008, and ICS 27 or ICS-27).
+- Don't overuse `code` format when writing in plain English.
+- Follow Google developer documentation [style guide](https://developers.google.com/style).
+- Check the meaning of words in Microsoft's [A-Z word list and term collections](https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/term-collections/accessibility-terms) (use the search input!).
+- We recommend using RFC keywords in user documentation (lowercase). The RFC keywords are: "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL. They are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).
+- Lint the markdown files for documentation with [markdownlint-cli](https://github.com/igorshubovych/markdownlint-cli). Run `make docs-lint` (you will need to have `markdownlint-cli` installed, so please follow the [installation instructions](https://github.com/igorshubovych/markdownlint-cli#installation)).
```
Comment (crodriguezvega) [inline docs/README.md:124]: I guess the reference to `markdownlint-cli` here will be updated with [this other PR](https://github.com/cosmos/ibc-go/pull/3805), right?

File: docs/architecture/adr-008-app-caller-cbs/transfer-callback-scaffold.md [hunk 5]
```diff
@@ -6,20 +6,20 @@ The exact nature of the callbacks to the smart contract will depend on the envir
...
 import (
- sdk "github.com/cosmos/cosmos-sdk/types"
- capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
-
- transfertypes "github.com/cosmos/ibc-go/v6/modules/apps/transfer/types"
- clienttypes "github.com/cosmos/ibc-go/v6/modules/core/02-client/types"
- channeltypes "github.com/cosmos/ibc-go/v6/modules/core/04-channel/types"
- porttypes "github.com/cosmos/ibc-go/v6/modules/core/05-port/types"
- "github.com/cosmos/ibc-go/v6/modules/core/exported"
+  sdk "github.com/cosmos/cosmos-sdk/types"
```
Comment (crodriguezvega) [inline docs/architecture/adr-008-app-caller-cbs/transfer-callback-scaffold.md:15]: Can the linter format the code blocks with 2-space tabs? If it could, it would make me sooooo happy.

File: docs/docs/01-ibc/03-apps/01-apps.md [hunk 6]
```diff
@@ -32,7 +32,7 @@ module correctly corresponding to the listed steps.
 
 :::note
 
-## Pre-requisites Readings
+## Pre-requisite readings
```
Comment (crodriguezvega) [inline docs/docs/01-ibc/03-apps/01-apps.md:35]: And can the linter make sure that headings use only capital letter on the first word? That would be like the best thing since sliced bread.

**Review by crodriguezvega — APPROVED** [review]
This is awesome, @srdtrk! I just have a couple of questions about the process for creating documentation for new releases, but I think I just need to do it once myself and it will be fine.

File: docs/docs/01-ibc/03-apps/01-apps.md (hunk 6 above)
Comment (damiannolan) [reply docs/docs/01-ibc/03-apps/01-apps.md:35]: I dunno man, sliced bread is pretty good if you ask me

File: docs/docs/05-migrations/06-v4-to-v5.md [hunk 7]
```diff
@@ -275,17 +276,17 @@ if err := proto.Unmarshal(ack.GetResult(), txMsgData); err != nil {
...
 default: // for SDK 0.45 and below
   for _, msgData := range txMsgData.Data {
-    // unmarshall msgData and execute logic based on the response 
+    // unmarshall msgData and execute logic based on the response
   }
 }
 ```
 
-See [ADR-03](../../architecture/adr-003-ics27-acknowledgement.md/#next-major-version-format) for more information or the [corrresponding documentation about authentication modules](../02-apps/01-interchain-accounts/03-auth-modules.md#onacknowledgementpacket).
+See [ADR-03](/architecture/adr-003-ics27-acknowledgement#next-major-version-format) for more information or the [corrresponding documentation about authentication modules](../02-apps/02-interchain-accounts/03-auth-modules.md#onacknowledgementpacket).
```
Comment (charleenfei) [inline docs/docs/05-migrations/06-v4-to-v5.md:289]:
```suggestion
See [ADR-03](/architecture/adr-003-ics27-acknowledgement#next-major-version-format) for more information or the [corresponding documentation about authentication modules](../02-apps/02-interchain-accounts/03-auth-modules.md#onacknowledgementpacket).
```

**Review by charleenfei — COMMENTED** [review]
first pass few nits

File: modules/apps/transfer/keeper/MBT_README.md [hunk 8]
```diff
@@ -46,4 +46,4 @@ To wrap Apalache docker image into an executable you might create the following
 docker run --rm -v $(pwd):/var/apalache apalache/mc $@
 ```
 
-In case of any questions please don't hesitate to contact Andrey Kuprianov (andrey@informal.systems).
+In case of any questions please don't hesitate to contact Andrey Kuprianov (<andrey@informal.systems>).
```
Comment (charleenfei) [inline modules/apps/transfer/keeper/MBT_README.md:49]: just wondering  if we still need these docs in here, not sure about latest state of Apalache?

**Review by charleenfei — APPROVED** [review]
thank you so much @srdtrk for taking care of this docs work, i know it's not the most fun task! 

i think we should go ahead and wrap up/merge this pr, and we can handle updating the versions ie 7.2.0 in a later pr.

File: modules/apps/transfer/keeper/MBT_README.md (hunk 8 above)
Comment (crodriguezvega) [reply modules/apps/transfer/keeper/MBT_README.md:49]: Apalache is still in [active development](https://github.com/informalsystems/apalache/commits/main), so I think it's ok to leave this here.

File: CODE_OF_CONDUCT.md [hunk 9]
```diff
@@ -34,7 +34,7 @@ This Code of Conduct applies both within project spaces and in public spaces whe
 
 ## Enforcement
 
-Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting the project team at community@interchain.io. The project team will review and investigate all complaints, and will respond in a way that it deems appropriate to the circumstances. The project team is obligated to maintain confidentiality with regard to the reporter of an incident. Further details of specific enforcement policies may be posted separately.
+Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting the project team at <community@interchain.io>. The project team will review and investigate all complaints, and will respond in a way that it deems appropriate to the circumstances. The project team is obligated to maintain confidentiality with regard to the reporter of an incident. Further details of specific enforcement policies may be posted separately.
```
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply CODE_OF_CONDUCT.md:37]: Checking!

**Comment by crodriguezvega** [comment]
Thank you, @srdtrk for taking of this migration, and @charleenfei for reviewing. Let's merge to unblock the rest of the work and open separate PRs for further improvements.
